src/lowering/desugar.ml

```diff
@@ -581,18 +581,20 @@ and obj obj_typ efs bases =
     base_dec, pick in
 
   let base_decs, pickers = map base_info bases |> split in
-  let gap T.{ lab; typ; _ } =
-    if exists (fun (ef : S.exp_field) -> ef.it.id.it = lab) efs then []
-    else
-      let id = fresh_var lab typ in
-      let [@warning "-8"] [base_var] = concat_map ((|>) lab) pickers in
-      let d =
-        if T.is_mut typ then
-          refD id { it = I.DotLE(varE base_var, lab); note = typ; at =  no_region }
-        else
-          letD id (dotE (varE base_var) lab typ) in
-      let f = { it = I.{ name = lab; var = id_of_var id }; at = no_region; note = typ } in
-      [d, f] in
+  let gap T.{ lab; typ; _ } = match typ with
+    | T.Typ _ -> []
+    | _ ->
+      if exists (fun (ef : S.exp_field) -> ef.it.id.it = lab) efs then []
+      else
+        let id = fresh_var lab typ in
+        let [@warning "-8"] [base_var] = concat_map ((|>) lab) pickers in
+        let d =
+          if T.is_mut typ then
+            refD id { it = I.DotLE(varE base_var, lab); note = typ; at =  no_region }
+          else
+            letD id (dotE (varE base_var) lab typ) in
+        let f = { it = I.{ name = lab; var = id_of_var id }; at = no_region; note = typ } in
+        [d, f] in
```

Comment:
You'll still need the type fields fs' to contains both type fields and value fields so I think you need to refine this a bit.
Maybe don't use map but a loop that accummulates the decs and type level fields separately, with different lengths.

Reply:
I guess you look ate a _late IR_, as directly after desugaring I see
``` Motoko
  (LetD
    (VarP ac)
    (BlockE
      (LetD (VarP $base/0) (VarE a))
      (LetD (VarP $base/1) (VarE c))
      (LetD (VarP $x/0) (PrimE (DotPrim x) (VarE $base/0)))
      (LetD (VarP $y/0) (PrimE (DotPrim y) (VarE $base/1)))
      (NewObjE
        Object
        (y $y/0)
        (x $x/0)
        {type T = Int; type X = Int; x : Int; y : () -> X}
      )
    )
  )
```

Reply:
Ah not, I misunderstood your code and thought fs were the field types, not the field values. Sorry. Looks good, I think.